python/ray/experimental/iter.py

```diff
@@ -0,0 +1,556 @@
+    tasks and actors. However, each shard should be read from at most one
+    process at a time.
+
+    Examples:
+        # Applying a function over items in parallel.
+        >>> it = ray.experimental.iter.from_items([1, 2, 3], num_shards=2)
+        ... <__main__.ParIterator object>
+        >>> it = it.for_each(lambda x: x * 2).gather_sync()
+        ... <__main__.LocalIterator object>
+        >>> print(list(it))
+        ... [2, 4, 6]
+
+        # Creating from generators.
+        >>> it = ray.experimental.iter.from_iterators([range(3), range(3)])
+        ... <__main__.ParIterator object>
+        >>> print(list(it.gather_sync()))
+        ... [0, 0, 1, 1, 2, 2]
+
+        # Accessing the individual shards of an iterator.
+        >>> it = ray.experimental.iter.from_range(10, num_shards=2)
+        ... <__main__.ParIterator object>
+        >>> it0 = it.get_shard(0)
+        ... <__main__.LocalIterator object>
+        >>> print(list(it0))
+        ... [0, 1, 2, 3, 4]
+        >>> it1 = it.get_shard(1)
+        ... <__main__.LocalIterator object>
+        >>> print(list(it1))
+        ... [5, 6, 7, 8, 9]
+
+        # Gathering results from actors synchronously in parallel.
+        >>> it = ray.experimental.iter.from_actors(workers)
+        ... <__main__.ParIterator object>
+        >>> it = it.batch_across_shards()
+        ... <__main__.LocalIterator object>
+        >>> print(next(it))
+        ... [worker_1_result_1, worker_2_result_1]
+        >>> print(next(it))
+        ... [worker_1_result_2, worker_2_result_2]
+    """
+
+    def __init__(self, actor_sets: List["_ActorSet"], name: str):
+        # We track multiple sets of actors to support parallel .union().
+        self.actor_sets = actor_sets
+        self.name = name
+
+    def __iter__(self):
+        raise TypeError(
+            "You must use it.gather_sync() or it.gather_async() to "
+            "iterate over the results of a ParIterator.")
+
+    def __str__(self):
+        return self.name
+
+    def __repr__(self):
+        return self.name
+
+    def for_each(self, fn: Callable[[T], T]) -> "ParIterator[T]":
+        """Remotely apply fn to each item in this iterator.
+
+        Arguments:
+            fn (func): function to apply to each item.
+
+        Examples:
+            >>> next(from_range(4).filter(lambda x: x * 2).gather_sync())
+            ... [0, 2, 4, 8]
+        """
+        return ParIterator(
+            [
+                a.with_transform(lambda local_it: local_it.for_each(fn))
+                for a in self.actor_sets
+            ],
+            name=self.name + ".for_each()")
+
+    def filter(self, fn: Callable[[T], bool]) -> "ParIterator[T]":
+        """Remotely filter items from this iterator.
+
+        Arguments:
+            fn (func): returns False for items to drop from the iterator.
+
+        Examples:
+            >>> next(from_items([0, 1, 2]).filter(lambda x: x).gather_sync())
```

Comment:
This example is a little bit confusing - `.filter(lambda x: x != 0)` or similar would be more obvious

Reply:
I actually prefer `lambda x: x`, though both seem fine. :)  
Interestingly you can actually just put `None` for the filter when filtering by truthiness... I'd imagine that's the most Pythonic way to do it, though it's probably confusing so maybe it's not the best option.